### PR TITLE
Refactor and clean up the decode/encode_pg_control functions

### PR DESCRIPTION
**Pull request by hlinnaka**
See commit messages.

This tries to document the transmute issues (https://github.com/zenithdb/zenith/issues/207) in code comments. How to solve that is a separate discussion, but let's at least add warnings to the code about it, until it's fixed.

### CODE REVIEW

File: postgres_ffi/build.rs [hunk 1]
```diff
@@ -11,16 +11,19 @@ fn main() {
     // to bindgen, and lets you build up options for
     // the resulting bindings.
     let bindings = bindgen::Builder::default()
-        // The input header we would like to generate
-        // bindings for.
+        // All the needed PostgreSQL headers are included from 'pg_control_ffi.h'
         .header("pg_control_ffi.h")
+
```
Comment (hlinnaka) [inline postgres_ffi/build.rs:16]: I added these newlines for clarity, to separate the different kinds of options. But `cargo fmt` insists on removing them again :-(. Any ideas?

**Comment by ericseppanen** [comment]
In #208 I was able to tidy up the crc code: [link](https://github.com/zenithdb/zenith/pull/208/files#diff-526683ea09888f69e04822fada89ba4d86b1bf4dd6c99e479d9e2af28b68c016R144). No `OFFSETOF_CRC` const needed.

Also, encoding and decoding functions should be implemented against the struct; no need for them to be free-floating functions.

File: postgres_ffi/build.rs (hunk 1 above)
Comment (funbringer) [reply postgres_ffi/build.rs:16]: Idk, what about this?

```
Builder::default()
    // section one
    .foo()
    .bar()
    .baz()
    //
    // section two
    .foo()
    .bar()
    .baz()
    //
    // section three
    .foo()
    .bar()
    .baz();
```

File: postgres_ffi/src/controlfile_utils.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    Ok(controlfile)
+}
+
+///
+/// Convert a struct representing a Postgres control file into raw bytes.
+///
+/// The CRC is recomputed to match the contents of the fields.
+pub fn encode_pg_control(controlfile: ControlFileData) -> Bytes {
```
Comment (funbringer) [inline postgres_ffi/src/controlfile_utils.rs:75]: I also feel that those two free functions belong to the `impl ControlFileData` above.

Then they would become

```rust
impl ControlFileData {
    pub fn decode(buf: &[u8]) -> anyhow::Result<Self>;  // becomes ControlFileData::decode at call site
    pub fn encode(&self) -> Bytes;  // e.g. ControlFileData::new().encode()
}
```

Note how we don't have to consume ControlFileData in encode.

File: postgres_ffi/src/controlfile_utils.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    //
+    // Use `transmute` to reinterpret struct as raw bytes.
+    //
+    // FIXME: This triggers undefined behavior, because the contents
```
Comment (funbringer) [inline postgres_ffi/src/controlfile_utils.rs:79]: Just wanted to say that I appreciate thoughtful comments like this one.

File: postgres_ffi/build.rs (hunk 1 above)
Comment (hlinnaka) [reply postgres_ffi/build.rs]: did it that way

**Comment by hlinnaka** [comment]
> In #208 I was able to tidy up the crc code: [link](https://github.com/zenithdb/zenith/pull/208/files#diff-526683ea09888f69e04822fada89ba4d86b1bf4dd6c99e479d9e2af28b68c016R144). No `OFFSETOF_CRC` const needed.

Ah, cool, I missed that part of #208. Cherry-picked that here.

**Comment by hlinnaka** [comment]
Thanks for the review so far. Can you please take another look now, this has changed quite a bit since the first version. Notably:

- I adopted the `memoffset::offset_of!` replacement for C offsetof()
- I turned encode/decode into method in ControlFileData
- I added a comment at the top of controlfile_utils.rs to give an introduction to what the control file is
- I removed the `ControlFileData::new()` constructor, it was unused

File: postgres_ffi/src/controlfile_utils.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        // undefined bytes into the resulting array. The Rust code won't
+        // care what's in those bytes, and PostgreSQL doesn't care
+        // either. HOWEVER, it is a potential security issue, because the
+        // bytes can contain arbitrary pieces of memory from the page
+        // server. In the worst case, that could be private keys or
+        // another tenant's data.
+        //
+        // See https://github.com/zenithdb/zenith/issues/207 for discussion.
+        let b: [u8; SIZEOF_CONTROLDATA] =
+            unsafe { std::mem::transmute::<ControlFileData, [u8; SIZEOF_CONTROLDATA]>(*self) };
```
Comment (funbringer) [inline postgres_ffi/src/controlfile_utils.rs:111]: Another way to do it:

```rust
let b: [u8; SIZEOF_CONTROLDATA] = unsafe {
    std::mem::transmute_copy(self)
};
```

Unfortunately, this function doesn't account for different src/dst sizes (which is unsafe squared), so I doubt that we should use it.

As an aside: rust is able to infer types in such cases, which means that we don't have to write the bulky `<ControlFileData, [u8; SIZEOF_CONTROLDATA]>`. Disregard this if you feel more comfortable with extra explicit types in unsafe blocks.

**Review by funbringer — APPROVED** [review]
LGTM, but I'd like to hear from Eric.

One more thing: I wonder if we should rename `encode -> to_bytes` and `decode -> from_bytes` to be more in line with [rust's naming conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv), but I have no strong opinions on this subject.

**Pull request merged**